subsystem/src/manager/manager_impl.rs

```diff
@@ -0,0 +1,301 @@
+
+        log::info!("Registering subsystem {full_name}");
+
+        let task = Box::pin(task::subsystem(
+            full_name.clone(),
+            subsys_init,
+            submit_handle.shallow_clone(),
+            action_rx,
+            shutdown_rx,
+            self.shutting_down_tx.clone(),
+        ));
+
+        self.subsystems.push(SubsystemData {
+            full_name,
+            task,
+            shutdown_tx,
+        });
+
+        Handle::new(submit_handle)
+    }
+
+    /// Add a subsystem that does not require custom initialization code.
+    pub fn add_subsystem<S>(&mut self, name: &'static str, subsys: S) -> Handle<S::Interface>
+    where
+        S: 'static + Send + Sync + Subsystem,
+    {
+        self.add_custom_subsystem(name, move |_| async {
+            Result::<S, std::convert::Infallible>::Ok(subsys)
+        })
+    }
+
+    /// Add a subsystem. Use the provided object directly as the subsystem state object.
+    pub fn add_direct_subsystem<S>(&mut self, name: &'static str, subsys: S) -> Handle<S>
+    where
+        S: 'static + Send + Sync,
+    {
+        self.add_subsystem(name, crate::wrappers::Direct::new(subsys))
+    }
+
+    /// Create a trigger object that can be used to shut down the system
+    pub fn make_shutdown_trigger(&self) -> ShutdownTrigger {
+        ShutdownTrigger::new(&self.shutting_down_tx)
+    }
+
+    /// Run the application main task.
+    ///
+    /// Completes when all the subsystems are fully shut down.
+    pub async fn main(mut self) {
+        let manager_name = self.config.name;
+        log::info!("Manager {manager_name} starting subsystems");
+
+        #[cfg(not(loom))]
+        let signal_handler_task = self.config.enable_signal_handlers.then({
+            let shutdown_trigger = self.make_shutdown_trigger();
+            || tokio::spawn(task::signal_handler(shutdown_trigger))
+        });
+
+        // Run all the subsystem tasks.
+        let subsystems: Vec<_> = self
+            .subsystems
+            .into_iter()
+            .map(|s| s.map_task(logging::spawn_in_current_span))
+            .collect();
+
+        // Signal the manager is shut down so it does not wait for itself
+        drop(self.shutting_down_tx);
+
+        // Wait for the shutdown trigger.
+        if self.shutting_down_rx.recv().await.is_none() {
+            log::warn!("Manager {manager_name}: all subsystems already down");
+        }
+        log::info!("Manager {manager_name} shutting down");
+
+        // Drop the receiver in order to prevent blocking of subsystems.
+        drop(self.shutting_down_rx);
+
+        // Shut down the subsystems in the reverse order of creation.
+        for subsys in subsystems.into_iter().rev() {
+            subsys.shutdown(self.config.shutdown_timeout_per_subsystem).await;
+        }
+
+        #[cfg(not(loom))]
+        if let Some(hdl) = signal_handler_task {
+            hdl.abort();
+            let _ = hdl.await;
+        }
```

Comment:
Not sure if this is relevant, but could any of the subsystems interact with the signal handler? If so, this block should probably be put above the shutdown for loop above.

Reply:
The shutdown signal may be delivered multiple times but there is no issue with it. That is a supported scenario because it's easily avoided in a concurrent system, e.g. you can issue a node shutdown RPC and hit Ctrl-C and both will issue the shutdown signal.

I'p pretty sure the shutdown procedure can be improved. For example, I don't think the Ctrl-C handler needs to be in a separate task at all. That requires further research. However, I think what we have now is correct.

Reply:
So I changed this so all the shutdown signals are rolled into a Future rather than having a separate task that listens to `Ctrl-C` and such. It makes the code somewhat simpler. It's in the last commit. Please take a look if that makes sense to you @alfiedotwtf.

Reply:
Much cleaner. I like it.